### ISSUE TITLE
"async & performance": ch4, Delegating Recursion

### ISSUE THREAD

**Issue by stefanbruvik** [getify/You-Dont-Know-JS#476]
I've been trying to wrap my brain around the recursive `yield`-delegation example in [ch4](https://github.com/getify/You-Dont-Know-JS/blob/master/async%20%26%20performance/ch4.md#delegating-recursion), but I can't quite seem to fully grok it, so I'm hoping you could clearify a couple of things:

In step 8, it says 

> When the promise resolves, its fulfillment message is sent [...] to the normal `yield` that's waiting in the `*foo(3)` generator instance.

But this is how I understand what's going on:
(I set up a working example over at jsbin: https://stebru.jsbin.com/noheco/2/edit?js,console, and according to the console output it seems to make sense the way I have (mis?-)understood this.)

When the first promise resolves, the response value (let's say its the number `1`) is sent (via `it.next(value)`in the `run()`method) to `*bar()`, which delegates to `*foo(3)`, which delegates to `*foo(2)`, which delegates to `*foo(1)`, where the waiting `return yield request(...);` is "replaced" with `return 1;`.

So the value is `return`ed (from `foo(1)`) to the waiting `val = yield *foo(1);` (in `foo(2)`), and so the yield expression and var assignment is replaced with `val = 1;`.

If my understanding is correct (which it probably isn't), step 8 should read:

> When the promise resolves, its fulfillment message is sent [...] to the normal `yield` that's waiting in the **`*foo(1)`** generator instance.

And in step 9 where it states:

> That first call's Ajax response is now immediately `return`ed from the `*foo(3)`generator instance [...]

it should rather read:

> That first call's Ajax response is now immediately `return`ed from the **`*foo(1)`** generator instance [...]

And lastly, in step 10 it states:

> [...] the second Ajax response propagates all the way back into the `*foo(2)`generator instance, and is assigned to its local `val` variable.

But as I understand it, the response is not assigned to the local `val` in `*foo(2)`, but rather replaced with the waiting `return yield request(...);`, e.g. `return 2`, and `return`èd from `*foo(2)` into the waiting `val = yield *foo(2);` in `*foo(3)` (which is replaced with `val = 2`).

I'm sure I'm just missing something here, but as my brain has now stopped working from thinking about all of this for too long, I hope you can help me out! :)

**Comment by getify** [comment]
First off, sorry for taking so incredibly long to respond. No excuse.

Secondly, I believe this is a duplicate of (or at least related to) the changes that I still plan to make to fix the confusion, per https://github.com/getify/You-Dont-Know-JS/pull/434.

**Comment by stebru** [comment]
Hah :) No worries, thanks for replying! I just hope my analysis is correct, though. (I thought I'd finally "got it"!!) But, yup, sure looks like a duplicate. Anyway, would be interested to hear if my reasoning is way off, or if I'm actually into something. 2nd edition, right? :)

As an aside:
This needs to be said: you really took JS to the next level! :) Thanks, Kyle! :)

**Comment by a-d-collins** [comment]
Yes, @getify. That you for uploading your book. I have found it to be very helpful thus far.

@stebru, I detected the same issue when I was reading through that section today. I am not an expert when it comes to generators-with-promises, but I'm fairly convinced you're right. Below is a .docx file that highlights the changes I think you're expecting in that section.

[Async_&_Performance-Delegating_Recursion.docx](https://github.com/getify/You-Dont-Know-JS/files/141350/Async_._Performance-Delegating_Recursion.docx)

**Comment by deepakcdo** [comment]
Yes I agree on this change too. This conufused me to. I really hope there was a branch with the know issues fixed . Even though its not perfect. Lots of people read the book and check master for errors. They might not check issues. 

This will lead to a major mis-understanding or giving up on reading the book as the errors are know but not easily available.

Especially considering the book was printed  in 2015 and its 2019 now :-)
